### PR TITLE
Publish merged realtime state as an atomic snapshot

### LINKED ISSUE #670: Performance: API Reader Starvation due to O(N) Map Allocation inside Write Lock (realTimeMutex)
### Description
Currently, `internal/gtfs/realtime.go` suffers from severe lock contention under load, leading to thread exhaustion and high p99 latency for API readers. 

The `rebuildMergedRealtimeLocked` function is called to rebuild the global routing maps (`realTimeTripLookup` and `realTimeVehicleLookupByTrip`). However, it performs this O(N) map allocation and data copying while holding an exclusive write lock (`realTimeMutex.Lock()`). 

### The Bottleneck
When processing a real-world GTFS-RT feed with tens of thousands of active trips, the map allocation takes several milliseconds. During this critical section:
1. Every incoming HTTP API request attempting to read state (e.g., calling `GetRealTimeTrips`) blocks while waiting for the `RLock`.
2. This creates a massive queue of blocked goroutines.
3. When the write lock is finally released, the "thundering herd" of blocked readers wakes up simultaneously, trashing the Go scheduler and causing cascading timeouts.

*Note: This architectural bottleneck perfectly aligns with the 75%+ failure rate and 1-minute latency spikes documented in `docs/mutex_contention_analysis.md`.*

### Proposed Solution: Lock-Free Copy-On-Write (COW)
To eliminate the reader starvation, we should move the O(N) allocation entirely out of the critical section using a Copy-On-Write pattern with `atomic.Value`.

**Implementation Steps:**
1. Group the lookup maps into a single state struct (e.g., `RealTimeState`).
2. Store this struct in the API manager using `atomic.Value`.
3. Inside the rebuild function, allocate and populate the new maps in local memory **without** acquiring the global lock.
4. Once the new maps are fully built, perform an O(1) atomic pointer swap to make them active.

This ensures that API readers never block waiting for background feed processing, keeping read latency strictly bounded to O(1) uncontended atomic loads.

@aaronbrethorst @Ahmedhossamdev
Should I work on this?

### PR DESCRIPTION
**Pull request by omlahore**
Fixes #670

`docs/mutex_contention_analysis.md` already names this one:

> **Bottleneck:** The `rebuildMergedRealtimeLocked` function holds a write lock while rebuilding large lookup maps, blocking all concurrent API readers.
>
> **Recommended Mitigation: Copy-On-Write (COW) with `atomic.Value`.** Readers will access it lock-free, while the 30s updater will swap in a new pre-computed version.

This does the reader half of that.

## The problem

`rebuildMergedRealtimeLocked` builds seven O(N) structures under `realTimeMutex.Lock()`: the merged trip and vehicle slices, three index maps, the duplicated-vehicle-by-route map, and a four-bucket alert index. Every API read path then took `RLock` to reach those fields. Go's `RWMutex` gives writers preference, so once a poll cycle starts a rebuild, incoming readers queue behind it.

## The change

The seven fields become one immutable `mergedRealtime`, published with `atomic.Pointer`. Readers call `manager.mergedRealtime()` and take no lock at all.

Writers still hold `realTimeMutex`. It continues to guard the per-feed maps the rebuild reads from, and the documented `staticMutex → realTimeMutex` ordering is untouched. **Writer hold time is unchanged.** This PR does not try to move the rebuild itself out of the lock, which needs a third mutex ordered against `realTimeMutex` and is really per-feed locking, #479.

## Measured

Reader latency sampled 3,000 times while rebuilds run in a tight loop:

| entities | | p50 | p99 | max |
|---|---|---|---|---|
| 10,000 | before | 1.128 ms | 3.093 ms | 7.361 ms |
| | after | **81 ns** | **471 ns** | **13.7 µs** |
| 50,000 | before | 7.345 ms | 16.625 ms | 29.949 ms |
| | after | **120 ns** | **992 ns** | **17.2 µs** |

Being straight about that benchmark: it rebuilds in a loop, which is far more aggressive than the real 30s poll, so the p50 column is not what production looks like. The honest reading is the tail. A reader that happens to arrive during a rebuild used to wait milliseconds and now waits for nothing, because it never takes the lock. I have deliberately not repeated the doc's "75%+ failure rate" figure, since I did not reproduce it.

## The awkward part: the mocks

`Mock*` helpers write the merged view directly instead of going through the feed maps, so they became copy-on-write: load, clone, mutate, store, still under `realTimeMutex` so concurrent mock writers serialise.

I deliberately did **not** "fix" them to route through `feedVehicles`/`feedTrips`. `vehicles_for_agency_handler_test.go:188` documents the current quirk in as many words, that `MockAddAlert` triggers a rebuild which wipes any directly injected vehicle, and roughly 170 mock call sites across the restapi tests sit on that behaviour. Changing it would turn this into a repo-wide test migration.

## Tests

`internal/gtfs/realtime_snapshot_test.go`:

- `TestMergedSnapshotIsImmutableAfterPublish` holds a snapshot across a rebuild and asserts it did not grow. This is the invariant the whole change rests on.
- `TestConcurrentReadsDuringRebuild` runs 8 readers against 40 rebuilds and checks every lookup index is consistent with the same snapshot's slice, so a torn read fails.
- `TestMergedRealtimeBeforeFirstPublish` covers the zero-value `Manager` that several tests construct, which now reads as empty instead of hitting nil maps.
- `BenchmarkRealtimeReadDuringRebuild` for the numbers above.

Checks, all clean:
- `go vet -tags "sqlite_fts5 sqlite_math_functions" ./...`
- `go vet -tags "purego" ./...`
- `go fmt ./...`
- `make test`, all 14 packages
- `go test -race` on `./internal/gtfs/` and `./internal/restapi/`

The race run matters more than usual here. Previously the readers were safe because they held `RLock`; now they are safe because the snapshot is never mutated after publish, and that is a claim worth having the detector check.

## Size and sequencing

366 insertions / 168 deletions across 12 files, which is over the 200-line guidance. Removing the fields does not compile without the mock and test migration in the same commit. Say the word if you would rather review it in two passes and I will split it into snapshot-plus-readers, then mocks-plus-tests.

I also updated `CLAUDE.md`, which still listed the merged fields as protected by `realTimeMutex` and mentioned a `realTimeAlerts` field that no longer exists, and added a status line to `docs/mutex_contention_analysis.md`.

One incidental cleanup: `GetVehicleForTrip` had a manual `RUnlock` on two branches rather than `defer`, which CONTRIBUTING warns against. The conversion deletes those locks entirely.

Note #1286 also edits `MockAddDuplicatedVehicle`. I will rebase onto it whenever it lands rather than race it.

## Summary by CodeRabbit

- **Improvements**
  - Improved real-time data access performance by enabling lock-free reads.
  - Real-time trips, vehicles, alerts, and lookup results are now delivered from consistent snapshots during updates.
  - Improved reliability for concurrent real-time updates and reads, reducing the risk of partially updated results.

- **Documentation**
  - Updated technical documentation to describe the real-time data handling and concurrency model.

### CODE REVIEW

File: internal/gtfs/realtime_snapshot_test.go [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			for !stop.Load() {
+				merged := manager.mergedRealtime()
+				// Every index in the lookup must be valid for the same
+				// snapshot's slice. A torn read would break this.
+				for id, idx := range merged.tripLookup {
+					require.Less(t, idx, len(merged.trips))
+					require.Equal(t, id, merged.trips[idx].ID.ID)
```
[the remark this answers is not in the document]
Comment (omlahore) [reply internal/gtfs/realtime_snapshot_test.go]: Fixed in c893a2ea. The workers now send the first mismatch on a buffered channel and the `require.NoError` happens after `wg.Wait()`, so `FailNow` only ever runs on the test goroutine.

File: internal/gtfs/realtime.go [hunk 2]
```diff
@@ -108,21 +183,15 @@ func (manager *Manager) cleanupExpiredVehicles(feedID string) {
...
-	defer manager.realTimeMutex.RUnlock()
-	return manager.realTimeTrips
+	return manager.mergedRealtime().trips
 }
 
 func (manager *Manager) GetRealTimeVehicles() []gtfs.Vehicle {
-	manager.realTimeMutex.RLock()
-	defer manager.realTimeMutex.RUnlock()
-	return manager.realTimeVehicles
+	return manager.mergedRealtime().vehicles
```
[the remark this answers is not in the document]
Comment (omlahore) [reply internal/gtfs/realtime.go:195]: Fixed the `require`-in-a-goroutine half in c893a2ea. On the deep copy I have pushed back rather than done it, so please overrule me if you disagree.

I checked every production caller before deciding. `getActiveTrips`, the `situationRefs` helpers, the `vehicles-for-agency` handler and the duplicated-vehicle loop in `trips_for_route_handler.go` all iterate and copy values out. Nothing mutates snapshot-owned data today, so this is a contract that is currently held rather than a live race.

Deep-copying every getter would undo what the PR is for. The snapshot exists so a read is a pointer load instead of a lock plus recomputation, and per-read deep copies put the allocation straight back, worse than the original since it would now happen on every call rather than every rebuild.

So I stated the read-only contract on the getters and added `TestConcurrentGetterReadsDuringRebuild`, which hammers the exported getters against 40 rebuilds under `-race` rather than only `mergedRealtime`. That pins the property the snapshot actually relies on. If you would rather have real immutability enforced, the cheaper version is returning cloned outer slices from the two slice getters so a caller cannot append or reorder into the shared array, and I would do that over full deep copies.

**Comment by aaronbrethorst** [comment]
### Code review

No issues found. Checked for bugs, CLAUDE.md compliance, and parity with the wiki spec and the Java OneBusAway server.

🤖 Generated with [Claude Code](https://claude.ai/code)

**Review by aaronbrethorst — APPROVED** [review]
Thanks. This is a careful implementation of the copy-on-write approach `docs/mutex_contention_analysis.md` recommended. I read the core of it:

- Each rebuild publishes a freshly allocated `mergedRealtime`.
- Every accessor loads the snapshot once and takes its indexes and slices from that one load.
- The `Mock*` helpers modify a copy under `realTimeMutex` instead of touching the published value.

The race detector is clean, and static endpoints on a King County Metro build match main.

The only red check is `TestBulkInsertPerformance` on Windows (5.07s against a 5s budget), which #1437 fixes and which this PR doesn't touch.

A few small things for a follow-up commit, none of them blocking:
- Several comments still say the code takes `realTimeMutex`: `internal/gtfs/realtime.go` around the alerts-for-trip doc, `internal/restapi/reference_utils.go`, and the "Acquire real-time lock independently" line in `gtfs_manager.go`.
- The CLAUDE.md edit says no read path acquires `realTimeMutex`, but `GetFeedUpdateTimes` still takes `RLock` (used by `/healthz` and metadata).
